internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+    Arr(Box<Array>),
+    Rec(Box<Row>),
+    Fun(Box<Function>),
+}
+
+impl fmt::Display for MonoType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MonoType::Bool => f.write_str("bool"),
+            MonoType::Int => f.write_str("int"),
+            MonoType::Uint => f.write_str("uint"),
+            MonoType::Flt => f.write_str("flt"),
+            MonoType::Str => f.write_str("str"),
+            MonoType::Dur => f.write_str("dur"),
+            MonoType::Time => f.write_str("time"),
+            MonoType::Rgx => f.write_str("rgx"),
+            MonoType::Var(var) => var.fmt(f),
+            MonoType::Arr(arr) => arr.fmt(f),
+            MonoType::Rec(obj) => obj.fmt(f),
+            MonoType::Fun(fun) => fun.fmt(f),
+        }
+    }
+}
+
+// Tvar stands for type variable.
+// A type variable holds an unknown type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Tvar(i64);
+
+impl fmt::Display for Tvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "t{}", self.0)
+    }
+}
+
+// Array is a homogeneous list type
+#[derive(Debug, Clone, PartialEq)]
+pub struct Array(MonoType);
+
+impl fmt::Display for Array {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[{}]", self.0)
+    }
+}
+
+// Row is an extensible record type.
+//
+// A row is either Nil meaning it has no properties,
+// or it is an extension of a row.
+//
+// A row may extend what is referred to as a row
+// variable. A row variable is a type variable that
+// represents an unknown record type.
+//
+#[derive(Debug, Clone)]
+pub enum Row {
+    Nil,
+    Var(Tvar),
+    Extension { head: Property, tail: Box<Row> },
+}
+
+impl fmt::Display for Row {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Row::Nil => f.write_str("{}"),
+            Row::Var(tv) => tv.fmt(f),
+            Row::Extension { head: h, tail: t } => write!(f, "{{{} | {}}}", h, t),
+        }
+    }
+}
+
+impl cmp::PartialEq for Row {
+    fn eq(&self, other: &Self) -> bool {
+        self.flatten() == other.flatten()
+    }
+}
+
+impl Row {
+    // Records are implemented as a sequence or list of extenstions.
+    // This function extends a record by adding a new property to the
+    // head of the list.
+    fn extend(self, head: Property) -> Self {
+        Row::Extension {
+            head: head,
+            tail: Box::new(self),
+        }
+    }
+
+    // Flatten a record type into a hashmap of property names and types
+    fn flatten(&self) -> (HashMap<String, &MonoType>, Option<Tvar>) {
+        match self {
+            Row::Nil => (HashMap::new(), None),
+            Row::Var(tv) => (HashMap::new(), Some(*tv)),
+            Row::Extension { head: h, tail: t } => {
+                let (mut r, tv) = t.flatten();
+                r.insert(h.k.clone(), &h.v);
+                (r, tv)
+            }
+        }
+    }
+}
```

Comment:
`flatten` is used only to compare two record types for equality. It should always be a private internal method.

Reply:
Makes sense, seems like later we should be able to implement an allocation free comparison that knows how to do the transitive lookup of properties within a row. But that can come later as an optimization.

Reply:
That doesn't seem necessary since this is just for testing. Perhaps this should be part of the test module instead of a struct method.